DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+    {
+        this.a = another;
+        this.b = "hello";
+    }
+}
+
+void main()
+{
+    C c;
+    C d = c;   // ok
+    A a;
+    C e = a;   // error
+}
+```
+2. It is illegal to declare a copy constructor for a struct that has a postblit defined and vice versa:
+
+```d
+struct A
+{
+    this(this) {}
+    @implicit this(ref A another) {}        // error, struct A defines a postblit
+}
+```
+
+Note that structs that do not define a postblit explicitly but contain fields that
+define one have a generated postblit. From the copy constructor perspective it makes
+no difference whether the postblit is user defined or generated:
+
+```d
+struct A
+{
+    this(this) {}
+}
+
+struct B
+{
+    A a;                                // A has a postblit -> __fieldPostblit is generated for B
+    @implicit this(ref B another) {}    // error, struct B defines a postblit
+}
+```
+
+#### Type checking
+
+The copy constructor typecheck is identical to the constructor one [[6](https://dlang.org/spec/struct.html#struct-constructor)]
+[[7](https://dlang.org/spec/struct.html#field-init)].
+
+The copy constructor overloads can be explicitly disabled:
+
+```d
+struct A
+{
+    @disable @implicit this(ref A another)
+    @implicit this(ref immutable A another)
+}
+
+void main()
+{
+    A b;
+    A a = b;     // error: disabled copy construction
+
+    immutable A ia;
+    A c = ia;  // ok
+
+}
+```
+
+In order to disable copy construction, all copy constructor overloads need to be disabled.
+In the above example, only copies from mutable to mutable are disabled; the overload for
+immutable to mutable copies is still callable.
+
+#### Overloading
+
+The copy constructor can be overloaded with different qualifiers applied to the parameter
+(copying from qualified source) or to the copy constructor itself (copying to qualified
+destination):
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}                        // 1 - mutable source, mutable destination
+    @implicit this(ref immutable A another) {}              // 2 - immutable source, mutable destination
+    @implicit this(ref A another) immutable {}              // 3 - mutable source, immutable destination
+    @implicit this(ref immutable A another) immutable {}    // 4 - immutable source, immutable destination
+}
+
+void main()
+{
+    A a;
+    immutable A ia;
+
+    A a2 = a;      // calls 1
+    A a3 = ia;     // calls 2
+    immutable A a4 = a;     // calls 3
+    immutable A a5 = ia;    // calls 4
+}
+```
+The proposed model enables the user to define the copying from an object of any qualified type
+to an object of any qualified type: any combination of 2 between mutable, `const`, `immutable`, `shared`,
+`const shared`.
+
+The `inout` qualifier may be used for the copy constructor parameter in order to specify that mutable, `const` or `immutable` types are
+treated the same:
+
+```d
+struct A
+{
+    @implicit this(ref inout A another) immutable {}
```

Comment:
Why does modifier `immutable` allow the assignment to mutable `a` below?

Reply:
This is indeed a mistake. The types are reversed : a,b,c should be immutable and the r's mutable, const, immutable